pydantic/utils.py

```diff
@@ -665,3 +677,39 @@ def all_identical(left: Iterable[Any], right: Iterable[Any]) -> bool:
         if left_item is not right_item:
             return False
     return True
+
+
+def get_discriminator_values(tp: Any, discriminator_key: str) -> Tuple[str, ...]:
+    """
+    Get all valid values in the `Literal` type of the discriminator field
+    `tp` can be a `BaseModel` class or directly an `Annotated` `Union` of many.
+    """
+    is_root_model = getattr(tp, '__custom_root_type__', False)
+
+    if get_origin(tp) is Annotated:
+        tp = get_args(tp)[0]
+
+    if hasattr(tp, '__pydantic_model__'):
+        tp = tp.__pydantic_model__
+
+    if is_root_model or is_union_origin(get_origin(tp)):
+        union_type = tp.__fields__[ROOT_KEY].type_ if is_root_model else tp
+
+        all_values = [get_discriminator_values(t, discriminator_key) for t in get_args(union_type)]
+        if len(set(all_values)) > 1:
+            raise TypeError(f'Field {discriminator_key!r} is not the same for all submodels of {tp.__name__!r}')
```

Comment:
While playing around, I get an `AttributeError` from `tp.__name__` with the code below.

I'm can imagine this code working, or raising this error, but we shouldn't get the attribute error.

```py
class Cat(BaseModel):
    pet_type: Literal['cat']
    name: str


class Dog(BaseModel):
    pet_type: Literal['dog']
    name: str


CommonPet = Annotated[Union[Cat, Dog], Field(discriminator='pet_type')]


class Lizard(BaseModel):
    pet_type: Literal['reptile', 'lizard']
    name: str


class Model(BaseModel):
    pet: Union[CommonPet, Lizard] = Field(..., discriminator='pet_type')
    n: int
```